packages/wrangler/src/api/startDevWorker/DevEnv.ts

```diff
@@ -36,70 +43,125 @@ export class DevEnv extends EventEmitter {
-		});
+		this.config = configFactory(this);
+		this.bundler = bundlerFactory(this);
+		this.runtimes = runtimeFactories.map((factory) => factory(this));
+		this.proxy = proxyFactory(this);
 
 		this.on("error", (event: ErrorEvent) => {
 			logger.debug(`Error in ${event.source}: ${event.reason}\n`, event.cause);
 			logger.debug("=> Error contextual data:", event.data);
 		});
+	}
 
-		config.on("configUpdate", (event) => {
-			bundler.onConfigUpdate(event);
-			proxy.onConfigUpdate(event);
-		});
-
-		bundler.on("bundleStart", (event) => {
-			proxy.onBundleStart(event);
-			runtimes.forEach((runtime) => {
-				runtime.onBundleStart(event);
-			});
-		});
-		bundler.on("bundleComplete", (event) => {
-			runtimes.forEach((runtime) => {
-				runtime.onBundleComplete(event);
-			});
-		});
-
-		runtimes.forEach((runtime) => {
-			runtime.on("reloadStart", (event) => {
-				proxy.onReloadStart(event);
-			});
-			runtime.on("reloadComplete", (event) => {
-				proxy.onReloadComplete(event);
-			});
-			runtime.on("devRegistryUpdate", (event) => {
-				config.onDevRegistryUpdate(event);
-			});
-		});
-
-		proxy.on("previewTokenExpired", (event) => {
-			runtimes.forEach((runtime) => {
-				runtime.onPreviewTokenExpired(event);
-			});
-		});
+	/**
+	 * Central message bus dispatch method.
+	 * All events from controllers flow through here, making the event routing explicit and traceable.
+	 *
+	 * Event flow:
+	 * - ConfigController emits configUpdate → BundlerController, ProxyController
+	 * - BundlerController emits bundleStart → ProxyController, RuntimeControllers
+	 * - BundlerController emits bundleComplete → RuntimeControllers
+	 * - RuntimeController emits reloadStart → ProxyController
+	 * - RuntimeController emits reloadComplete → ProxyController
+	 * - RuntimeController emits devRegistryUpdate → ConfigController
+	 * - ProxyController emits previewTokenExpired → RuntimeControllers
+	 * - Any controller emits error → DevEnv error handler
+	 */
+	dispatch(event: ControllerEvent): void {
+		switch (event.type) {
+			case "error":
+				this.handleErrorEvent(event);
+				break;
+
+			case "configUpdate":
+				this.bundler.onConfigUpdate(event);
+				this.proxy.onConfigUpdate(event);
+				break;
+
+			case "bundleStart":
+				this.proxy.onBundleStart(event);
+				this.runtimes.forEach((runtime) => {
+					runtime.onBundleStart(event);
+				});
+				break;
+
+			case "bundleComplete":
+				this.runtimes.forEach((runtime) => {
+					runtime.onBundleComplete(event);
+				});
+				break;
+
+			case "reloadStart":
+				this.proxy.onReloadStart(event);
+				break;
+
+			case "reloadComplete":
+				this.proxy.onReloadComplete(event);
+				break;
+
+			case "devRegistryUpdate":
+				this.config.onDevRegistryUpdate(event);
+				break;
+
+			case "previewTokenExpired":
+				this.runtimes.forEach((runtime) => {
+					runtime.onPreviewTokenExpired(event);
+				});
+				break;
+
+			default: {
+				const _exhaustive: never = event;
```

Comment:
There's an `assertNever()` helper somewhere in the codebase which we can use here